### PR TITLE
Ensure that controls update all their sizing information when required

### LINKED ISSUE #74052: Label + [autowrap] in VBoxContainer changes its size at startup (like line_spacing)
### Godot version

v4 RC3 - RC6

### System information

Windows 10

### Issue description

If a Label with any autowrap enabled is in the VBoxContainer, then depending on the type of autowrap, the Label size changes along the Y axis when it is first displayed (or the project is launched). As if at this moment the line_spacing parameter increases. If you change line_spacing through the code after the program is launched (the project file is attached), then everything immediately falls into place, and in all the Labels of this container.
I understand that this is not a great problem and it can be circumvented with one line of code, however, I spent several hours trying to figure out why my Popup window does not take the correct size at the time of the first call, and why everything falls into place after the second call. It turns out I had a VBoxContainer with Labels there, which behaved like this.
PS: changing the size of the main window, apparently, leads to a forced redrawing of the Labels, and also everything falls into place.
label_test

### Steps to reproduce

image
The autowrap parameter for Labels should be different from OFF

### Minimal reproduction project

The project is attached. By clicking on the buttons or changing the screen size, you can see this issue.
[_labels_test.zip](https://github.com/godotengine/godot/files/10840488/_labels_test.zip)

### LINKED ISSUE #77787: Script name tooltips are oversized
### Godot version

4.1 db5113d

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 (NVIDIA; 30.0.15.1403) - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 Threads)

### Issue description

image
Notice the empty space to the right. I actually have a worse case in my project, but I couldn't reproduce it outside:

Regression from #77651
CC @Rindbee 

### Steps to reproduce

1. Add a new script
2. Name it `;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;.gd`
3. Hover its name in the Script Editor

### Minimal reproduction project

N/A

### LINKED ISSUE #77815: 4.1 SubViewport.size is always (2, 2)?
### Godot version

v4.1.dev.custom_build [543750a1b]

### System information

Windows 11, Core i5-12400F, RX 6600

### Issue description

	extends SubViewport

	func _ready():
		print(size)

- in 4.0.3- this prints the resolution of the window `(1152, 648)`
- in 4.1 this prints `(2, 2)`

Thanks.

### Steps to reproduce

- new `Control` scene
  - add `SubViewportContainer` child
    - set `Anchor` to `Full Rect`
    - turn `Stretch` on
    - add `SubViewport` child
      - add `Camera2D` child
          - set `Anchor Mode` to `Fixed Top Left`
      - attach script that prints `size`

### Minimal reproduction project

https://github.com/rakkarage/TestSubViewportSize

### LINKED ISSUE #77885: Container size not updating correctly when using call_deferred
### Godot version

543750a1b3f5696f9ba8e91cb49dc7db05d2ae62

### System information

Windows 10, Vulkan

### Issue description

This is a regression since v4.1.dev4.official [5c2295ff5], I can't bisect atm to find the commit that caused it, maybe https://github.com/godotengine/godot/pull/77651 but that's just a guess.
When a HBoxContainer has a label child, and that label's text is changed, one used to be able to find the new size of the container by using call_deferred, this is no longer the case.

MRP correct behavior (v4.1.dev4.official [5c2295ff5]):
image
MRP incorrect (543750a1b3f5696f9ba8e91cb49dc7db05d2ae62):
image

### Steps to reproduce

```gdscript
extends Control

@onready var h_box_container: HBoxContainer = $HBoxContainer
@onready var label: Label = $HBoxContainer/Label

func _ready() -> void:
	label.text = "larger non default text"
	print("h_box_container size.x: ", h_box_container.size.x)
	call_deferred("size_after_text_changed")

func size_after_text_changed() -> void:
	print("deferred h_box_container size.x: ", h_box_container.size.x)
```

### Minimal reproduction project

[call_deferred bug.zip](https://github.com/godotengine/godot/files/11654441/call_deferred.bug.zip)

### LINKED ISSUE #78091: Atlas preview jumps after first scroll event
### Godot version

4.1 beta1

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 (NVIDIA; 30.0.15.1403) - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 Threads)

### Issue description

When I change tile source, the view is initialized at a wrong position. After scrolling it for the first time, the position gets re-adjusted.

https://github.com/godotengine/godot/assets/2223172/830cef4a-10f0-4fa5-9172-e7f8015330eb

It wasn't happening before.

### Steps to reproduce

1. Create a TileMap with TileSet
2. Add a sufficiently big tile atlas source
3. Add another source, so that you can switch between the 2
4. Switch to the other source and then to the big one
5. Scroll

### Minimal reproduction project

The texture from the video: https://github.com/godotengine/godot/assets/2223172/7ac78271-84d0-48ab-ae28-45841e2e0880

### LINKED ISSUE #78118: 4.1.beta1, mouse-over slider shows offset handle highlight
### Godot version

v4.1.beta1.official [828ec2c5d]

### System information

Godot v4.1.beta1 - Windows 10.0.19045 - Vulkan (Compatibility) - GeForce GTX 550 Ti () - Intel(R) Core(TM) i3-2100 CPU @ 3.10GHz (4 Threads)

### Issue description

Issue observed in:
- v4.1.beta1.official [828ec2c5d]

Expected behavior in:
- v4.1.dev4.official [5c2295ff5]
- v4.0.stable.official [92bee43ad]
- v3.5.2.stable.official [170ba337a]

### Description

A sliders mouse-over handle highlight (a circle with a ring around it) appears offset (down and to the right), the first time the mouse is moved over a slider.

offset

### Expected
The slider handle highlight, should appear on the sliders handle on the first mouse-over.
As it does in the Godot versions before: 4.1.beta1

expected

### Steps to reproduce

- open the attached project: [issue, mouse over slider, handle highlight, offset.zip](https://github.com/godotengine/godot/files/11714943/issue.mouse.over.slider.handle.highlight.offset.zip)
Node2D is selected, in the Scene panel, and the Transform group/section is unfolded in the Inspector panel.
- move the mouse over the Transform > Rotation, slider

### Observed

The sliders mouse-over handle highlight (a circle with a ring around it) appears offset (down and to the right) from the sliders handle.

The mouse-over handle highlight moves to the slider handle when:
- the mouse is held still over the slider until the tooltip appears
- or if the mouse is moved away from the slider and back over the slider

Now the second and subsequent slider mouse-overs keep showing the handle highlight correctly.

The issue can be reproduced by deselecting and selecting Node2D in the Scene panel.

https://github.com/godotengine/godot/assets/13420573/c817788c-6032-44e0-8693-35801f406c51

### Expected

The sliders mouse-over handle highlight should appear on the sliders handle, the first time the mouse is moved over the slider.

### Minimal reproduction project

[issue, mouse over slider, handle highlight, offset.zip](https://github.com/godotengine/godot/files/11714943/issue.mouse.over.slider.handle.highlight.offset.zip)

### PR DESCRIPTION
**Pull request by YuriSizov**
This PR reverts https://github.com/godotengine/godot/pull/77651 and supersedes https://github.com/godotengine/godot/pull/77825.

It still fixes https://github.com/godotengine/godot/issues/74052, the original issue from https://github.com/godotengine/godot/pull/77651. It also fixes regressions from that PR:
- Fixes https://github.com/godotengine/godot/issues/77815.
- Fixes https://github.com/godotengine/godot/issues/77885.
- Fixes https://github.com/godotengine/godot/issues/77787 (though I couldn't reproduce it even with the current master, so please double-check).
- Fixes #78118 (tested by Sauermann)
- Fixes https://github.com/godotengine/godot/issues/78091

I decided to take the approach of reverting and refixing the issue because I wasn't confident in the original solution. I had my doubts initially, but it looked sensible enough and fixed the issue at hand, so I approved and merged it. However, given all the new issues that popped up, I thought it would be better for me to investigate the problem further and try a safer approach.

My idea here is that instead of changing the order in which the data is invalidated, we should instead make sure that all the necessary routines are called when we think that the size might've changed. This happens on tree entry, but also on global changes, such as themes and translations. Previously we would only do some of the updates in these situations, but I'm not sure it was appropriate to do it like that.

I have a suspicion, that for the most part it worked thanks to a very nasty hidden side-effect of `get_combined_minimum_size()`, which could in turn trigger a complete invalidation of the minimum size information. Yes, from an innocent getter. This wasn't originally the case, however, this was bolted on later, in https://github.com/godotengine/godot/pull/19871 (a follow up to https://github.com/godotengine/godot/pull/19334). I don't think this was a proper solution, but once again it solved the problems at hand at the time. I removed that side-effect, and with all the other changes haven't noticed any obvious regressions, testing the editor and the test scenes from all the linked issues.

This brings things closer to the original state of this commit: https://github.com/godotengine/godot/commit/005b69cf6e276209464cc8c36ebc7376679925b6#diff-2b7ba034670859ced66e442b9ec59f7351febc3b4352e535bc85da804bcca266

-----

I was faced with one regression of my own, where opening a scene in the editor in the background caused size calculations to misbehave once you switched to it. Turns out that when we request an update with `_update_minimum_size()` we set a guard flag to prevent doing it multiple times. However, when the node is out of the scene tree, this method hits an early return. Which means that the guard flag is never reset and the method cannot be triggered again either.

For some reason, this only happened in editor and only under those specific circumstances (instanced scenes weren't affected, for example). Changing the size of the control in the editor would also cause it to finally update, somehow. So it's still a bit of a mystery, but the logic seems flawed without this fix.

Edit: Well, I guess for this situation to happen, the call to `update_minimum_size` must happen while the nodes are inside of the tree, and then the call to `_update_minimum_size` (with the underscore, the actual routine) needs to happen while the nodes are out of the tree. Which is... interesting? I guess as scenes are opened in the editor, the nodes are briefly added to the tree and then removed?

-----

All in all, a lot of the original issue comes down to the old problem of cascading updates in labels (see https://github.com/godotengine/godot/pull/73809). Basically, when we do shaping of a text label, we use its current size. But shaping itself affects the minimum size, which in turn further affects the current size. This means that shaping is an unstable and unreliable operation whenever we have wrapping enabled (and thus cannot enforce any minimum size based on the contents).

This probably affects more stuff than labels, but somehow the entire GUI system manages to work to this day, and this PR seems to address the particular issue with auto-wrapping by doing just enough updates to make things right.

### CODE REVIEW

**Comment by akien-mga** [comment]
CC @Rindbee

**Comment by Rindbee** [comment]
The reason for this series of issues is `Label`.

1. Issue a `NOTIFICATION_RESIZED` notification (`_size_changed()`); https://github.com/godotengine/godot/blob/e188d619227990001667821dac8bc8940076d4a9/scene/gui/control.cpp#L1736
2. Mark `data.minimum_size_valid` as `false`; https://github.com/godotengine/godot/blob/e188d619227990001667821dac8bc8940076d4a9/scene/gui/control.cpp#L1589

Then calling `get_combined_minimum_size()` may **immediately** update `data.minimum_size_cache`. And the update of `data.last_minimum_size` is **deferred** in `_update_minimum_size()`. This mechanism has a long history and is widely used.

The potential problem is that `data.minimum_size_cache` has been updated multiple times and then returns to `data.last_minimum_size`, so the following code will not be executed. https://github.com/godotengine/godot/blob/e188d619227990001667821dac8bc8940076d4a9/scene/gui/control.cpp#L1609-L1613

Instead, the parent container was previously using an outdated minimum size of child label to calculate its own minimum size. 

So for labels, when `data.minimum_size_cache` returns to `data.last_minimum_size`, if `data.minimum_size_cache` has been updated, a `minimum_size_changed` signal may need to be emitted to notify the parent container to update the minimum size.

File: scene/gui/control.cpp [hunk 1]
```diff
@@ -1579,12 +1579,28 @@ Vector2 Control::get_pivot_offset() const {
...
 	if (!is_inside_tree()) {
+		data.updating_last_minimum_size = false;
 		return;
 	}
 
-	Control *invalidate = this;
+	Size2 minsize = get_combined_minimum_size();
+	data.updating_last_minimum_size = false;
+
+	if (minsize != data.last_minimum_size) {
```
Comment (Rindbee) [inline scene/gui/control.cpp:1589]:
```suggestion
	if (minsize != data.last_minimum_size || data.minimum_size_cache_update_ahead ) {
	data.minimum_size_cache_update_ahead = false;
```

File: scene/gui/control.cpp [hunk 2]
```diff
@@ -1665,17 +1665,8 @@ void Control::_update_minimum_size_cache() {
 	minsize.x = MAX(minsize.x, data.custom_minimum_size.x);
 	minsize.y = MAX(minsize.y, data.custom_minimum_size.y);
 
-	bool size_changed = false;
-	if (data.minimum_size_cache != minsize) {
-		size_changed = true;
-	}
-
 	data.minimum_size_cache = minsize;
```
Comment (Rindbee) [inline scene/gui/control.cpp:1668]:
```suggestion
	if (data.minimum_size_cache != minsize) {
		data.minimum_size_cache_update_ahead = true;
	}
	data.minimum_size_cache = minsize;
```

**Comment by YuriSizov** [comment]
> I'm not sure if this PR can avoid the above situation.

Can you provide a test case for that situation? It doesn't matter how many times the `data.minimum_size_cache` is invalidated, because it will get updated the moment it is requested next. And its value is the source of truth and it is to be trusted. So parent containers always receive the actual and correct information about children min sizes.

As for the trigger for the min size changed signal, it is the responsibility of each control to call `update_minimum_size` whenever it is expected that the min size was updated. And when that happens we will check what the current source of truth holds (or computes on the fly) and compare it with the stored value of `data.last_minimum_size`. This is the only place where `data.last_minimum_size` can be set, so I don't see how there can be any issue with that comparison. It will be reliably hit as long as the control calls `update_minimum_size` when appropriate.

**Comment by Rindbee** [comment]
> Can you provide a test case for that situation? 

I can't provide a test case for this PR. But I'm sure #74052 has this kind of problem. See https://github.com/godotengine/godot/issues/74052#issuecomment-1572048492

**Comment by YuriSizov** [comment]
The underlying problem is that auto-wrapping makes label sizing unstable, and we cannot fix that with a patch like that. It also affects other controls that behave similarly, like the FlowContainer.

When a control is designed to be elastic and adapt to whatever outer size restrictions there are, but does not have any internal restrictions, its size cannot be resolved in one step like we do for all other controls. In this case the Label cannot enforce any minimum size because it has wrapping enabled, so its minimum size is 0. This means that until the actual size is enforced by the container the Label will shape its text as if the actual width was 0, increasing the height to fit all the text according to its wrapping rules.

Then when the actual size is established the Label will shape the text closer to what expected and reset its minimum size requirements to a normal height. And after that the container needs to do the sorting again.

It always takes several steps. The current GUI system is designed to propagate minimum size changes upwards and reach stability in one such pass. But with wrapping there is this back and forth where parent containers affect the minimum size of their children, and thus a second pass is needed. And as a result we get various bugs, where the second pass doesn't happen or we get cascading changes snowballing and not stopping.

I wasn't trying to fix the underlying problem, because it's not a simple fix. Fixing would require us to redesign some things about our GUI system. What I was trying to do was to make sure we react to resizing in more straightforward and predictable manner, without partial or hidden logic, and it happened to fix the issue without the same regressions your solution has.

**Comment by Rindbee** [comment]
> The underlying problem is that auto-wrapping makes label sizing unstable

This is a more underlying reason.

https://github.com/godotengine/godot/blob/a5a23678c4e9bb4e137caf7db1bec8900d91cef3/scene/gui/control.cpp#L1607

The above code, viewed from another perspective, is equivalent to the following code:

```cpp
	get_combined_minimum_size(); // Treat it as an update request.
	Size2 minsize = data.minimum_size_cache;
```

So the work of `_update_minimum_size()` can be regarded as synchronizing `data.minimum_size_cache` to `data.last_minimum_size`.

**Comment by YuriSizov** [comment]
> And the following code is somewhat equivalent to executing part of the `_update_minimum_size()` code immediately (just not emitting the `minimum_size_changed` signal)

The difference here is that calling `_size_changed()` explicitly alongside `update_minimum_size()` means that we will double-check if the actual size has changed, and then, the next time the `MessageQueue` is flushed, we will check if the min size has changed, and if it did, check the actual size again. In the case of `Label` this also means that we would likely request a reshaping immediately, and then once again if the min size change, which seems to bring the balance to the whole thing.

But anyway, I'm sorry, I'm not sure if we're discussing semantics at this point or if you're pointing at some flaw? Can you elaborate if there is something that you believe I'm missing? So far everything seems to work without changing the overall logic.

**Review by KoBeWi — APPROVED** [review]
I didn't spot regressions from these changes and the PR fixes multiple issues, so it's fine.

**Comment by Sosasees** [comment]
¿will Godot with this pull request merged
update Control sizes when changing the _window size_ project settings?

so far i had to repress the anchor preset on the canvas's topmost Control after changing the _window size_ project setting to preview how the new window size would really look like.

(this is the only reason i added FullRectControl to my [_responsive UI plugin_ version 1.1](https://codeberg.org/sosasees/godot-4-responsive-ui-plugin/src/commit/ced9d13e9fab3469a71efcd8033035ab570b58b9).
if the question hears yes, i can remove it in the add-on's possible version 2.0)

**Pull request merged**